docs/design/jet/023-jobupload.md

```diff
@@ -0,0 +1,147 @@
+- The client should send multi parts in a sequential manner. Out of order messages are not handled since it would require more resources on the server.
+  
+
+### User Interaction
+#### API design and/or Prototypes
+
+```java
+HazelcastInstance client = ...;
+JetService jetService = client.getJet();
+List<String> jobParameters = emptyList();
+
+String job1 = "job1";
+//If there is an error, throws JetException
+jetService.uploadJob(getJarPath(),
+        null,
+        job1,
+        null,
+        jobParameters);
+```
+
+#### Client Related Changes
+A new method has been added to JetService interface. Any client that implements this interface, has to implement this method
+```java
+void uploadJob(@Nonnull Path jarPath, String snapshotName, String jobName, String mainClass, List<String> jobParameters);
+```
+
+### Technical Design
+
+The client protocol needs to support job uploading. So that non-java clients can also upload jet jobs.
+For this purpose two new messages are added to client protocol
+
+1. uploadJobMetaData
+2. uploadJobMultipart
+
+**1. uploadJobMetaData Message**
+
+The upload process starts with uploadJobMetaData. This message contains
+
+| Term                   | Definition                                                       |
+|------------------------|------------------------------------------------------------------|
+| sessionId | The UUID. This field associates all messages in a session |
+| jarSize | The jar size in bytes                                            |
+| snapshotName | Argument passed when starting the job                            |
+| jobName | Argument passed when starting the job                            |
+| mainClass | Argument passed when starting the job                            |
+| jobParameters | Argument passed when starting the job                            |
+
+
+Upon reception of uploadJobMetaData message, the server performs some validation rules. If any of them fail, a **JetException** is thrown. If all is well, server stores a new entry in the JobUploadStore class
+
+**2. uploadJobMultipart message**
+
+The upload process continues with this message. It contains jar's bytes. This message contains these fields
+
+| Term                   | Definition                                                              |
+|------------------------|-------------------------------------------------------------------------|
+| sessionId | It is explained in the previous message |
+| currentPartNumber | It starts from 1 and shows the sequence number of the part. For example 1 of 5 |
+| totalPartNumber | It is the total number of parts of the sequence  |
+| partData | is the **byte[]** containing jar data  |
```

Comment:
Should we also include a checksum for this part data?

Reply:
If we are going to do checksum for the whole jar, then there is no need for checksum of parts